### ISSUE TITLE
SC1039: Indented heredoc pattern should not be considered content

### ISSUE THREAD

**Issue by ingydotnet** [koalaman/shellcheck#1198]
#### For bugs
- Rule Id: SC1039
- My shellcheck version: 0.4.7
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
cat <<FOO
  foo
    bar
  FOO
baz
FOO
```

#### Here's what shellcheck currently says:
```
In foo.bash line 5:
  FOO
  ^-- SC1039: Use <<- instead of << if you want to indent the end token.
```

#### Here's what I wanted or expected to see:
Should not get an error.

**Comment by matthewpersico** [comment]
Yeah, but it's still valid:
```bash
matthew@monolith:~$ cat x
#!/bin/bash

cat <<FOO
  foo
    bar
  FOO
baz
FOO

cat <<'  FOO'
  foo
    bar
  FOO
baz
FOO
matthew@monolith:~$ ./x
  foo
    bar
  FOO
baz
  foo
    bar
./x: line 14: baz: command not found
./x: line 15: FOO: command not found
matthew@monolith:~$
```

**Comment by Dmole** [comment]
Did shell check say otherwise or was it just warning you about ugly code? (the -f gcc option lists the type)

**Comment by ingydotnet** [comment]
@Dmole, this was just an example to show the bug. The point of heredocs is to quote *any* text, without escaping anything, as long as that text does not have a line containing only the ending token.

That means you can have entire programs (even bash scripts) as heredocs, and they can contain quoted strings and even more heredocs. Or you might have documentation or XML or all sorts of things. In other words, quotes are no substitute for the elegant power of heredocs.

I've actually hit this bug in real code. It's a pretty bad UX, because shellcheck finds the wrong end and then usually gets confused by the text immediately following and gives up.

**Comment by matthewpersico** [comment]
Ha. Look at my comment with the code above - even the markdown highlighter gets it wrong.

**Comment by koalaman** [comment]
This is related to #1050. ShellCheck does intend to warn for this code, because it is a relatively common error. The bug is that you can't just disable the warning, because it's parsed as stopping on that line. I.e., this perfectly valid script fails to parse because of it:

```
#!/bin/bash
cat <<FOO
  FOO
()
FOO
```

This was done intentionally at the time, but like most such "helpful" parsing choices, it turned out to be a bad idea.
